add --shared option for activate

Evizero opened the pull request:
This is my first naive/minimal attempt to add a `--shared` option to `activate`.
The idea here is that one could use this to have some sort of global environment at a default path.

```
julia> Pkg.pkg"activate --shared Foo"

(Foo) pkg> status
    Status `~/.julia/enviroments/Foo/Project.toml`
```

Is this something we would like to pursue further? Its a feature that I personally would like to have, and I'd be happy to do my best to develop it further with a little guidance. I tried to mirror how `dev` is handling `--shared` and `--local`

00vareladavid commented on src/Pkg.jl:
```diff
@@ -7,6 +7,7 @@ using REPL.TerminalMenus
 depots() = Base.DEPOT_PATH
 logdir() = joinpath(depots()[1], "logs")
 devdir() = get(ENV, "JULIA_PKG_DEVDIR", joinpath(depots()[1], "dev"))
+envdir() = joinpath(depots()[1], "enviroments")
```
I think this is a typo. Should it be "environments"?

00vareladavid commented on src/API.jl:
```diff
@@ -589,6 +589,8 @@ function activate(path::Union{String,Nothing}=nothing)
         else
             path = abspath(path)
         end
+    elseif path !== nothing && !envdir
```
It makes more sense to me if `envdir == true` is when you use ` envdir()`

00vareladavid replied:
Ah, I see. It is a detail anyway.

fredrikekre commented on src/REPLMode.jl:
```diff
@@ -1104,7 +1104,10 @@ packages have changed causing the current Manifest to_indices be out of sync.
-    [],
+    [
+        ("local", OPT_SWITCH, :envdir => true),
+        ("shared", OPT_SWITCH, :envdir => false),
```
I think the kwargs here are better as `:local => true` and `:local => false`.

Evizero replied:
I don't think I can use `local` because its is a keyword that I can't use as a kw argument idnetifier. I'll try `shared` instead

fredrikekre commented:
So basically,
```
pkg> activate --shared Foo
```
is just a shorthand for
```
pkg> activate $(joinpath(depots()[1], "enviroments"))/Foo
```
which I am a bit sceptical to, the `--shared` option does not do anything different, just appends a path. Compare with `dev` where `--shared` means "put this in some shared location, and `--local` means "put this in the local project". But maybe this is useful anyway? I don't know, I just feel like environments are not shared or local in the same sense.

Evizero commented:
Allow me to maybe sketch out the workflow I have in mind with this.

Its pretty typical that if I work on the JuliaImages ecosystem (or the JuliaML ecosystem), that I have multiple packages there checked out which I work on in sync. For example to move functionality out of Images.jl to ImageTransformations.jl

The setup/workflow that would feel convenient here is:

- the `v0.7` environment is on the tagged versions of everything for normal work
- I have a `JuliaImages` environment where all the JuliaImages packages are marked as `dev` (and similarily a `JuliaML` environment for the JuliaML ones etc)

1.) I cd to the concrete package I want to work on in `.julia/dev` (say `ImageCore`).
2.) Aside from VIM, I usually want to have a REPL start in that project dir so I can do `include("test/runtest.jl")`.
3.) It would be convenient if when I start a REPL there, that I can say `activate --shared JuliaImages`, which in my setup would be shared for all those packages.

One alternative would be that each JuliaImage package is its own environment/project where i always do `activate .`, but that seems a bit redundant, since then I have to specify which other JuliaImages packages are on `dev` for each of them individually.

Evizero commented:
> So basically,
>
> ```
> pkg> activate --shared Foo
> ```
>
> is just a shorthand for
>
> ```
> pkg> activate $(joinpath(depots()[1], "enviroments"))/Foo
> ```

Right. Maybe with the added restriction that in such a case the `path` is only allowed to be a valid directory name (so not "Foo/bar", or "/Foo")

> I like this because it is a good mental transition between "environments as paths" to "environments as symbolic names". It makes the REPL more user friendly.

I'd like to think so.

> If you do not want the API to be affected(which I agree is a valid concern), the logic can always be moved to `REPLMode.do_activate()`.

I could try that. though currently it also seems unintrusive with the simple `shared` keyword 
>

fredrikekre commented:
> I like this because it is a good mental transition between "environments as paths" to "environments as symbolic names".

Yea that makes sense. So in `pkg> activate --shared Foo`
1. Should we require `Foo` to exist?
2. Should we look in all depots for `Foo`?

Maybe something like: Look for `Foo` in all depots, if we don't find it, activate/create `depots[1]/Foo`?

Evizero commented:
> 1. Should we require Foo to exist?

I am torn on this. on the one hand one wants to catch typos, on the other hand I'd like to avoid having to `cd` there manually first to make the directory. Maybe log and `info` in the case the environment doesn't exist yet?  In the current implementation the folder is only created after oder does any other operation. So in the case the user had a typo and realized his/her mistake, and then proceeded to `activate --shared Foo` the existing one, then no harm would be done.

> 2. Should we look in all depots for Foo?

That is a great idea

Evizero commented:
Ok so I implemented the heuristic you suggested and also added the info message in case the named environment does not yet exist

```
julia> Pkg.pkg"activate --shared Foo"
[ Info: New shared environment "Foo" will be placed at /Users/csto/.julia/environments/Foo
```

Or should we go without the info? If one does `activate Foo` for a non-existing Foo no message is printed either. so this would be a bit inconsistent

00vareladavid commented on src/API.jl:
```diff
@@ -570,26 +570,43 @@ function status(ctx::Context, mode=PKGMODE_PROJECT)
     else
-        path = abspath(path)
+        # this regex disallows names such as "Foo/bar", ".", "..", "Foo*", etc
+        occursin(r"([\\/:*?<>|])|(^\.+$)", path) && cmderror("not a valid name for a shared environment: $(path)")
```
I'm against using regex here because regex is hard to read and it is difficult to determine all the edge cases across the different OS we support. I would prefer if we first determine depot and then:
```
basename(joinpath(depot,path)) == path # error out if this is false
```
I think it reflects the intention more directly.

Evizero replied:
That wouldn't catch `.` or `..` though. but maybe that's not the responsibility of this API

Evizero replied:
or I could do the following, which does catch `.` and `..`

```
basename(abspath(joinpath(fullpath,path))) == path
```

I'll try it

KristofferC commented:
I don't understand how an enviornment can be "shared". The `--shared` flag for devving packages is because multiple environments will end up *sharing* the same package but nothing is sharing the enviornment this generates?

00vareladavid commented:
Perhaps `shared` is a bad name? The main utility I see from this PR is when users don't know or don't care about the path to an environment. Perhaps something like `--canonical` would be better?

Evizero commented:
The motivation to use `--shared` was because it already exists in another context that was relateable in my eyes (i.e without shared `dev` puts it locally where you are, while with shared it puts it in a subfolder of your main depot). We could call it `--named` or `--global` instead if its about the wording.

Do you see my usecase/argument for the ability to have such named environments?

StefanKarpinski commented:
I think `--shared` is a good name. `--global` could work too but I think it's more confusing.

Evizero commented:
I incorporated the feedback. I'll write some tests tomorrow. Maybe this is also a good time to add a docstring to `activate` since it doesn't have one yet

Evizero commented:
how about `--depot` ?

Evizero commented on src/API.jl:
```diff
@@ -569,27 +569,47 @@ function status(ctx::Context, mode=PKGMODE_PROJECT)
+            isdir(fullpath) && break
+        end
+        # this disallows names such as "Foo/bar", ".", "..", etc
+        if basename(abspath(joinpath(fullpath,path))) != path
```
oh right thats an embarrassing mistake. also good point about line 608. Thanks!

00vareladavid replied:
Welcome!

Evizero replied:
moving the line to 608 would allow the user to say `activate --shared .` (which would resolve to the `environments` folder itself, as well as any `../../xx` to some existing folder. Unsure if it makes sense to allow or disallow that. At the very least `activate --shared .` seems like something that no one would every conciously want to do, so this might be a good case where we should rather throw some error

KristofferC commented:
Ok, I can buy the argument that using `--shared` everywhere.

Evizero commented:
While writing tests for it, I realized that having a `--local` option to denote "not shared" would be a bit strange for the `activate` command. Here is why: One would be tempted to assume that `activate --local Foo` would always be equivalent to `activate ./Foo` (i.e. sidestepping a devved Foo if one exists). 

I am a bit unsure what to do here, so for now I removed the `--local` option

fredrikekre commented:
Yea, I don't think we need to have the `--local` flag here.

Evizero commented:
I think this is ready unless someone has any more comments / change requests

fredrikekre commented on src/API.jl:
```diff
@@ -568,27 +568,47 @@ function status(ctx::Context, mode=PKGMODE_PROJECT)
+        # unless the shared environment already exists, place it in the first depots
+        if !isdir(fullpath)
+            fullpath = joinpath(Pkg.envdir(Pkg.depots1()), path)
+            @info("new shared environment \"$path\" will be placed at $fullpath")
```
Should we warn here? We dont warn for `pkg> activate NonExistingDirectory`. I prefer not warning here I think.

Evizero replied:
well i actually changed both to print an info for consistency. I could change it back though if thats prefered. My reasoning for printing an info is to catch typos more easily

00vareladavid replied:
I prefer the `@info` warning, for the same reason as @Evizero. I don't think being too spammy will be an issue here, since creating a new environment should happen less often then activating an existing one.

fredrikekre replied:
1. I think you will notice a typo, since you will notice that there is nothing there?
2. If we keep the `@info`, may I suggest that we print something alone the lines of 
```
[Info: activating new [shared] environment $path in $fullpath.
```

Evizero replied:
Do we want to print `$path` even for non-shared? since then it could very well be something like `../Foo/Bar` which isn't much of a environment name

fredrikekre replied:
We could print the full, normalized path for `../Foo/Bar`.

00vareladavid replied:
In my experience, subtle issues like this are easy to miss when buried under other issues (or when you are a new user of an interface). Then you go: "Argh! I can't believe THAT was the issue!" and throw your computer out the window.

:+1:  @fredrikekre's ` @info` message

KristofferC replied:
We can tweak the exact way to info things in the upcoming PR but I aim to get the new commits into the rc3 so I merged this quickly.

00vareladavid replied:
Yes we can later do a pass and make all the infos/warnings consistent with each other.

Evizero commented:
I am confused @KristofferC I did update the string that you are linking. Is there something about it I should change further?

KristofferC commented:
No, I just missed it somehow :D

Evizero commented:
Thanks everyone for all the feedback and patience

StefanKarpinski commented:
Thanks for ge contribution! Always great to get more people who can make changes to the package manager.
